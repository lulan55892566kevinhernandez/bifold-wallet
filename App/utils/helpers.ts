import { ConnectionRecord, CredentialRecord, ProofRecord, RequestedAttribute } from '@aries-framework/core'
import { useConnectionById, useCredentialById, useProofById } from '@aries-framework/react-hooks'
import startCase from 'lodash.startcase'
import { parseUrl } from 'query-string'

import { indyCredentialKey, IndexedIndyCredentialMetadata } from '../constants'

export function parseSchema(schemaId?: string): { name: string; version: string } {
  let name = 'Credential'
  let version = ''
  if (schemaId) {
    const schemaIdRegex = /(.*?):([0-9]):([a-zA-Z .\-_0-9]+):([a-z0-9._-]+)$/
    const schemaIdParts = schemaId.match(schemaIdRegex)
    if (schemaIdParts?.length === 5) {
      name = `${schemaIdParts?.[3].replace(/_|-/g, ' ')}`
        .split(' ')
        .map((schemaIdPart) => schemaIdPart.charAt(0).toUpperCase() + schemaIdPart.substring(1))
        .join(' ')
      version = schemaIdParts?.[4]
    }
  }
  return { name, version }
}

export function credentialSchema(credential: CredentialRecord): string | undefined {
  return credential.metadata.get<IndexedIndyCredentialMetadata>(indyCredentialKey)?.schemaId
}

export function parsedSchema(credential: CredentialRecord): { name: string; version: string } {
  return parseSchema(credentialSchema(credential))
}

/**
 * Adapted from https://stackoverflow.com/questions/3426404/create-a-hexadecimal-colour-based-on-a-string-with-javascript
 */
export function hashCode(s: string): number {
  return s.split('').reduce((hash, char) => char.charCodeAt(0) + ((hash << 5) - hash), 0)
}

export function hashToRGBA(i: number) {
  const colour = (i & 0x00ffffff).toString(16).toUpperCase()
  return '#' + '00000'.substring(0, 6 - colour.length) + colour
}

export function credentialRecordFromId(credentialId?: string): CredentialRecord | void {
  if (credentialId) {
    return useCredentialById(credentialId)
  }
}

export function connectionRecordFromId(connectionId?: string): ConnectionRecord | void {
  if (connectionId) {
    return useConnectionById(connectionId)
  }
}

export function proofRecordFromId(proofId?: string): ProofRecord | void {
  if (proofId) {
    return useProofById(proofId)
  }
}

export function getConnectionName(connection: ConnectionRecord | void): string | void {
  if (!connection) {
    return
  }
  return connection?.alias || connection?.invitation?.label
}

export function firstMatchingCredentialAttributeValue(attributeName: string, attributes: RequestedAttribute[]): string {
  if (!attributes.length) {
    return ''
  }
  const firstMatchingCredential = attributes[0].credentialInfo
  if (!firstMatchingCredential) {
    return ''
  }
  const match = Object.entries(firstMatchingCredential.attributes).find(
    ([n]) => startCase(n) === startCase(attributeName)
  )
  return match?.length ? match[1] : ''
}

export const isRedirection = (url: string): boolean => {
  const queryParams = parseUrl(url).query
  return !(queryParams['c_i'] || queryParams['d_m'])
}
