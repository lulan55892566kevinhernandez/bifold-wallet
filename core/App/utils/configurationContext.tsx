import { createContext, useContext } from 'react'

import { pages } from '../screens/OnboardingPages'

export interface ConfigurationContext {
  onboarding: {
    pages: typeof pages
  }
  splash: React.FC
  terms: React.FC
}

const ConfigurationContext = createContext<ConfigurationContext>(null as unknown as ConfigurationContext)
export const ConfigurationProvider = ConfigurationContext.Provider

export const useConfigurationContext = () => useContext(ConfigurationContext)
