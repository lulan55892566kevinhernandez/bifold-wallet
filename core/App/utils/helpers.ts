import {
  Agent,
  ConnectionRecord,
  CredentialExchangeRecord,
  RequestedAttribute,
  RequestedPredicate,
  RetrievedCredentials,
  IndyCredentialInfo,
  IndyProofFormat,
} from '@aries-framework/core'
import {
  FormatDataMessagePayload,
  FormatRetrievedCredentialOptions,
} from '@aries-framework/core/build/modules/proofs/models/ProofServiceOptions'
import { useConnectionById } from '@aries-framework/react-hooks'
import { Buffer } from 'buffer'
import moment from 'moment'
import { parseUrl } from 'query-string'

import { i18n } from '../localization/index'
import { ProofCredentialAttributes, ProofCredentialPredicates } from '../types/record'

export { parsedCredDefName } from './cred-def'
export { parsedSchema } from './schema'

/**
 * Generates a numerical hash based on a given string
 * @see https://stackoverflow.com/questions/3426404/create-a-hexadecimal-colour-based-on-a-string-with-javascript
 * @param { string } s given string
 * @returns { number } numerical hash value
 */
export const hashCode = (s: string): number => {
  return s.split('').reduce((hash, char) => char.charCodeAt(0) + ((hash << 5) - hash), 0)
}

/**
 * Generates a pseudorandom number between 0 and 1 based on a seed
 * @see https://gist.github.com/tommyettinger/46a874533244883189143505d203312c
 * @see https://stackoverflow.com/questions/521295/seeding-the-random-number-generator-in-javascript
 * @param { number } seed any number
 * @returns { number } pseudorandom number between 0 and 1
 */
const mulberry32 = (seed: number) => {
  let t = (seed += 0x6d2b79f5)
  t = Math.imul(t ^ (t >>> 15), t | 1)
  t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
  return ((t ^ (t >>> 14)) >>> 0) / 4294967296
}

/**
 * Converts a numerical hash into a hexidecimal color string
 * @see https://helderesteves.com/generating-random-colors-js/#Generating_random_dark_colors
 * @param { number } hash numerical hash value (generated by hashCode function above)
 * @returns { string } hexidecimal string eg. #32d3cc
 */
export const hashToRGBA = (hash: number) => {
  let color = '#'
  const colorRangeUpperBound = 256

  // once for r, g, b
  for (let i = 0; i < 3; i++) {
    // append a pseudorandom two-char hexidecimal value from the lower half of the color spectrum (to limit to darker colors)
    color += ('0' + Math.floor((mulberry32(hash + i) * colorRangeUpperBound) / 2).toString(16)).slice(-2)
  }

  return color
}

export function formatTime(time: Date, params?: { long?: boolean; format?: string }): string {
  const getMonthKey = 'MMMM'
  const momentTime = moment(time)
  const monthKey = momentTime.format(getMonthKey)
  const customMonthFormatRe = /M+/
  const long = params?.long
  const format = params?.format
  const shortDateFormatMaskLength = 3

  let formatString = i18n.t('Date.ShortFormat')
  if (format) {
    formatString = format
  } else {
    if (long) {
      formatString = i18n.t('Date.LongFormat')
    }

    // if translation fails
    if (formatString === 'Date.ShortFormat' || formatString === 'Date.LongFormat' || formatString === undefined) {
      formatString = 'MMM D, YYYY'
    }
  }
  const customMonthFormat = formatString?.match(customMonthFormatRe)?.[0]

  let formattedTime = momentTime.format(formatString)

  if (customMonthFormat) {
    let monthReplacement = ''
    const monthReplacementKey = momentTime.format(customMonthFormat)
    if (customMonthFormat.length === shortDateFormatMaskLength) {
      // then we know we're dealing with a short date format: 'MMM'
      monthReplacement = i18n.t(`Date.MonthShort.${monthKey}`)
    } else if (customMonthFormat.length > shortDateFormatMaskLength) {
      // then we know we're working with a long date format: 'MMMM'
      monthReplacement = i18n.t(`Date.MonthLong.${monthKey}`)
    }
    // if translation doesn't work
    if (monthReplacement === `Date.MonthLong.${monthKey}` || monthReplacement === `Date.MonthShort.${monthKey}`) {
      monthReplacement = monthReplacementKey
    }

    if (monthReplacement) {
      formattedTime = formattedTime.replace(monthReplacementKey, monthReplacement)
    }
  }
  return formattedTime
}

/**
 * @deprecated The function should not be used
 */
export function connectionRecordFromId(connectionId?: string): ConnectionRecord | void {
  if (connectionId) {
    return useConnectionById(connectionId)
  }
}

/**
 * @deprecated The function should not be used
 */
export function getConnectionName(connection: ConnectionRecord | void): string | void {
  if (!connection) {
    return
  }
  return connection?.alias || connection?.theirLabel
}

export function getCredentialConnectionLabel(credential?: CredentialExchangeRecord) {
  if (!credential) {
    return ''
  }

  return credential?.connectionId
    ? useConnectionById(credential.connectionId)?.theirLabel
    : credential?.connectionId ?? ''
}

export function getConnectionImageUrl(connectionId: string) {
  const connection = useConnectionById(connectionId)
  if (!connection) {
    return undefined
  }
  return connection.imageUrl ?? undefined
}

export function firstValidCredential(
  fields: RequestedAttribute[] | RequestedPredicate[],
  revoked = true
): RequestedAttribute | RequestedPredicate | null {
  if (!fields.length) {
    return null
  }

  let first = null
  const firstNonRevoked = fields.filter((field) => !field.revoked)[0]
  if (firstNonRevoked) {
    first = firstNonRevoked
  } else if (fields.length && revoked) {
    first = fields[0]
  }

  if (!first?.credentialInfo) {
    return null
  }

  return first
}

export const getOobDeepLink = async (url: string, agent: Agent | undefined): Promise<any> => {
  const queryParams = parseUrl(url).query
  const b64Message = queryParams['d_m'] ?? queryParams['c_i']
  const rawmessage = Buffer.from(b64Message as string, 'base64').toString()
  const message = JSON.parse(rawmessage)
  await agent?.receiveMessage(message)
  return message
}

/**
 * A sorting function for the Array `sort()` function
 * @param a First retrieved credential
 * @param b Second retrieved credential
 */
export const credentialSortFn = (a: any, b: any) => {
  if (a.revoked && !b.revoked) {
    return 1
  } else if (!a.revoked && b.revoked) {
    return -1
  } else {
    return b.timestamp - a.timestamp
  }
}

export const processProofAttributes = (
  request?: FormatDataMessagePayload<[IndyProofFormat], 'request'> | undefined,
  credentials?: FormatRetrievedCredentialOptions<[IndyProofFormat]>
): { [key: string]: ProofCredentialAttributes } => {
  const processedAttributes = {} as { [key: string]: ProofCredentialAttributes }

  if (!(request?.indy?.requested_attributes && credentials?.proofFormats?.indy?.requestedAttributes)) {
    return {}
  }

  const requestedProofAttributes = request.indy.requested_attributes
  const retrievedCredentialAttributes = credentials.proofFormats.indy.requestedAttributes

  for (const key of Object.keys(retrievedCredentialAttributes)) {
    // The shift operation modifies the original input array, therefore make a copy
    const credential = [...(retrievedCredentialAttributes[key] ?? [])].sort(credentialSortFn).shift()

    const credName = key
    let revoked = false
    if (credential) {
      revoked = credential.revoked as boolean
    }
    const { name, names } = requestedProofAttributes[key]

    for (const attributeName of [...(names ?? (name && [name]) ?? [])]) {
      if (!processedAttributes[credName]) {
        // init processedAttributes object
        processedAttributes[credName] = {
          schemaId: credential?.credentialInfo?.schemaId,
          credDefId: credential?.credentialInfo?.credentialDefinitionId,
          credName,
          attributes: [],
        }
      }

      let attributeValue = '' //(credentialInfo as IndyCredentialInfo).attributes[attributeName]
      if (credential) {
        attributeValue = (credential.credentialInfo as IndyCredentialInfo).attributes[attributeName]
      }
      processedAttributes[credName].attributes?.push({
        revoked,
        name: attributeName,
        value: attributeValue,
      })
    }
  }
  return processedAttributes
}

export const processProofPredicates = (
  request?: FormatDataMessagePayload<[IndyProofFormat], 'request'> | undefined,
  credentials?: FormatRetrievedCredentialOptions<[IndyProofFormat]>
): { [key: string]: ProofCredentialPredicates } => {
  const processedPredicates = {} as { [key: string]: ProofCredentialPredicates }

  if (!(request?.indy?.requested_predicates && credentials?.proofFormats?.indy?.requestedPredicates)) {
    return {}
  }

  const requestedProofPredicates = request.indy.requested_predicates
  const retrievedCredentialPredicates = credentials.proofFormats.indy.requestedPredicates

  for (const key of Object.keys(requestedProofPredicates)) {
    // The shift operation modifies the original input array, therefore make a copy
    const credential = [...(retrievedCredentialPredicates[key] ?? [])].sort(credentialSortFn).shift()
    const { credentialId, revoked, credentialDefinitionId, schemaId } = { ...credential, ...credential?.credentialInfo }
    const { name, p_type: pType, p_value: pValue } = requestedProofPredicates[key]

    const credName = key

    if (!processedPredicates[credName]) {
      processedPredicates[credName] = {
        schemaId,
        credDefId: credentialDefinitionId,
        credName: credName,
        predicates: [],
      }
    }

    processedPredicates[credName].predicates?.push({
      credentialId,
      name,
      revoked,
      pValue,
      pType,
    })
  }
  return processedPredicates
}

/**
 * @deprecated The function should not be used
 */
export const sortCredentialsForAutoSelect = (credentials: RetrievedCredentials): RetrievedCredentials => {
  const requestedAttributes = Object.values(credentials?.requestedAttributes).pop()
  const requestedPredicates = Object.values(credentials?.requestedPredicates).pop()
  const sortFn = (a: any, b: any) => {
    if (a.revoked && !b.revoked) {
      return 1
    } else if (!a.revoked && b.revoked) {
      return -1
    } else {
      return b.timestamp - a.timestamp
    }
  }

  requestedAttributes && requestedAttributes.sort(sortFn)
  requestedPredicates && requestedPredicates.sort(sortFn)

  return credentials
}

/**
 *
 * @param url a redirection URL to retrieve a payload for an invite
 * @param agent an Agent instance
 * @returns payload from following the redirection
 */
export const receiveMessageFromUrlRedirect = async (url: string, agent: Agent | undefined) => {
  const res = await fetch(url, {
    method: 'GET',
    headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
  })
  const message = await res.json()
  await agent?.receiveMessage(message)
  return message
}

/**
 *
 * @param url a redirection URL to retrieve a payload for an invite
 * @param agent an Agent instance
 * @returns payload from following the redirection
 */
export const receiveMessageFromDeepLink = async (url: string, agent: Agent | undefined) => {
  const res = await fetch(url, {
    method: 'GET',
    headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
  })
  const message = await res.json()
  await agent?.receiveMessage(message)
  return message
}

/**
 *
 * @param uri a URI containing a base64 encoded connection invite in the query parameter
 * @param agent an Agent instance
 * @returns a connection record from parsing and receiving the invitation
 */
export const connectFromInvitation = async (uri: string, agent: Agent | undefined) => {
  const invitation = await agent?.oob.parseInvitation(uri)

  if (!invitation) {
    throw new Error('Could not parse invitation from URL')
  }

  const record = await agent?.oob.receiveInvitation(invitation)
  const connectionRecord = record?.connectionRecord
  if (!connectionRecord?.id) {
    throw new Error('Connection does not have an ID')
  }

  return connectionRecord
}
