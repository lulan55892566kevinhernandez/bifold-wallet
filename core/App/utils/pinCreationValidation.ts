import { PINRules } from '../types/security'

const consecutiveSeriesOfThree = new RegExp(/012|123|234|345|456|567|678|789|987|876|765|654|543|432|321|210/)
const evenNumberSeries = new RegExp('(13579)')
const oddNumberSeries = new RegExp('(02468)')
const isNumber = new RegExp('^[0-9]+$')
const crossNumberPattern = ['159753', '159357', '951357', '951753', '357159', '357951', '753159', '753951']

export interface PinValidationsType {
  isInvalid: boolean
  errorName:
    | 'CrossPatternValidation'
    | 'OddOrEvenSequenceValidation'
    | 'NoRepetitionOfTheSameNumbersValidation'
    | 'NoRepetitionOfTheTwoSameNumbersValidation'
    | 'NoSeriesOfNumbersValidation'
    | 'PinOnlyContainDigitsValidation'
    | 'PinTooShortValidation'
    | 'PinTooLongValidation'
}

export const pinCreationValidations = (pin: string, pinRules: PINRules) => {
  const pinValidations: PinValidationsType[] = []
  if (pinRules.no_cross_pattern) {
    pinValidations.push({
      isInvalid: crossNumberPattern.includes(pin),
      errorName: 'CrossPatternValidation',
    } as PinValidationsType)
  }
  if (pinRules.no_even_or_odd_series_of_numbers) {
    pinValidations.push({
      isInvalid: evenNumberSeries.test(pin) || oddNumberSeries.test(pin),
      errorName: 'OddOrEvenSequenceValidation',
    } as PinValidationsType)
  }
  if (pinRules.no_repeated_numbers) {
    let noRepeatedNumbers = new RegExp(/(\d)\1{1,}/)
    if (typeof pinRules.no_repeated_numbers === 'number') {
      noRepeatedNumbers = new RegExp(`(\\d)\\1{${pinRules.no_repeated_numbers - 1},}`)
    }
    pinValidations.push({
      isInvalid: noRepeatedNumbers.test(pin),
      errorName: 'NoRepetitionOfTheSameNumbersValidation',
    } as PinValidationsType)
  }
  if (pinRules.no_repetition_of_the_two_same_numbers) {
    let noRepetitionOfTheTwoSameNumbers = new RegExp(/([0-9][0-9])\1{1,}/)
    if (typeof pinRules.no_repetition_of_the_two_same_numbers === 'number') {
      noRepetitionOfTheTwoSameNumbers = new RegExp(
        `([0-9][0-9])\\1{${pinRules.no_repetition_of_the_two_same_numbers - 1},}`
      )
    }
    pinValidations.push({
      isInvalid: noRepetitionOfTheTwoSameNumbers.test(pin),
      errorName: 'NoRepetitionOfTheTwoSameNumbersValidation',
    } as PinValidationsType)
  }
  if (pinRules.no_series_of_numbers) {
    pinValidations.push({
      isInvalid: consecutiveSeriesOfThree.test(pin),
      errorName: 'NoSeriesOfNumbersValidation',
    } as PinValidationsType)
  }
  if (pinRules.only_numbers) {
    pinValidations.push({
      isInvalid: !isNumber.test(pin),
      errorName: 'PinOnlyContainDigitsValidation',
    } as PinValidationsType)
  }

  pinValidations.push({
    isInvalid: pin.length < pinRules.min_length || pin.length > pinRules.max_length,
    errorName: pin.length < pinRules.max_length ? 'PinTooShortValidation' : 'PinTooLongValidation',
  } as PinValidationsType)

  return pinValidations
}
