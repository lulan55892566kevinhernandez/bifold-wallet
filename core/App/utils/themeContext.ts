import { createContext, useContext } from 'react'

import { defaultTheme, Theme } from '../theme'

export { defaultTheme } from '../theme'
export type { Theme } from '../theme'
export const ThemeContext = createContext<Theme>(defaultTheme)
export const ThemeProvider = ThemeContext.Provider

export const useThemeContext = () => useContext(ThemeContext)
