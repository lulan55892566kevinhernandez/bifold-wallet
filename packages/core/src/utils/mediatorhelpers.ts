import { Agent, MediationRecord } from '@credo-ts/core'

export const isMediatorInvitation = async (agent: Agent, url: string): Promise<boolean> => {
  const invitation = await agent.oob.parseInvitation(url)
  if (!invitation) {
    return false
  }
  if (invitation.goalCode === 'aries.vc.mediate') {
    agent.config.logger.info(`Invitation is a mediator invitation with goal code: ${invitation.goalCode}`)
    return true
  }
  agent.config.logger.info(`Invitation is not a mediator invitation, goal code: ${invitation.goalCode}`)
  return false
}

const provisionMediationRecordFromMediatorUrl = async (
  agent: Agent,
  url: string
): Promise<MediationRecord | undefined> => {
  try {
    const invitation = await agent.oob.parseInvitation(url)
    if (!invitation) {
      agent.config.logger.warn(`No invitation found in URL: ${url}`)
      return undefined
    }
    const outOfBandRecord = await agent.oob.findByReceivedInvitationId(invitation.id)
    let [connection] = outOfBandRecord ? await agent.connections.findAllByOutOfBandId(outOfBandRecord.id) : []

    if (!connection) {
      agent.config.logger.warn(`No connection found for out-of-band record: ${outOfBandRecord?.id}`)
      const invite = await agent.oob.parseInvitation(url)
      const { connectionRecord: newConnection } = await agent.oob.receiveInvitation(invite)
      if (!newConnection) {
        agent.config.logger.warn(`Failed to create connection from invitation: ${JSON.stringify(invite, null, 2)}`)
        return
      }
      connection = newConnection
    }
    const result = connection.isReady ? connection : await agent.connections.returnWhenIsConnected(connection.id)
    return agent.mediationRecipient.provision(result)
  } catch (error) {
    agent.config.logger.warn(`Failed to get connection ID from mediator URL: ${error}`)
    return
  }
}

export const setMediationToDefault = async (agent: Agent, mediatorUrl: string) => {
  const mediationRecord = await provisionMediationRecordFromMediatorUrl(agent, mediatorUrl)
  if (!mediationRecord) {
    agent.config.logger.warn(`No connection record found for mediator URL: ${mediatorUrl}`)
    return
  }
  const currentDefault = await agent.mediationRecipient.findDefaultMediator()
  if (currentDefault?.connectionId === mediationRecord.id) {
    agent.config.logger.info(`Default mediator already set for connection ID: ${mediationRecord.id}`)
    return
  }
  await agent.mediationRecipient.setDefaultMediator(mediationRecord)
  agent.config.logger.info(`setting default mediator with record: ${JSON.stringify(mediationRecord)}`)
}
