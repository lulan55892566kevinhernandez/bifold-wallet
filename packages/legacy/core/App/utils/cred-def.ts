import { CredentialMetadataKeys, CredentialExchangeRecord as CredentialRecord } from '@aries-framework/core'

import { credentialSchema, parseSchemaFromId } from './schema'

export function parseCredDefFromId(credDefId?: string, schemaId?: string): string {
  let name = 'Credential'
  if (credDefId) {
    const credDefRegex = /[^:]+/g
    const credDefIdParts = credDefId.match(credDefRegex)
    if (credDefIdParts?.length === 5) {
      name = `${credDefIdParts?.[4].replace(/_|-/g, ' ')}`
        .split(' ')
        .map((credIdPart) => credIdPart.charAt(0).toUpperCase() + credIdPart.substring(1))
        .join(' ')
    }
  }
  if (name.toLocaleLowerCase() === 'default' || name.toLowerCase() === 'credential') {
    name = parseSchemaFromId(schemaId).name
  }
  return name
}

function credentialDefinition(credential: CredentialRecord): string | undefined {
  return credential.metadata.get(CredentialMetadataKeys.IndyCredential)?.credentialDefinitionId
}

export function parsedCredDefName(credential: CredentialRecord): string {
  return parseCredDefFromId(credentialDefinition(credential), credentialSchema(credential))
}
