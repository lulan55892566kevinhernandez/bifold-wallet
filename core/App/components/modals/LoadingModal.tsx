import React from 'react'
import { ActivityIndicator, Dimensions, Image, Modal, SafeAreaView, StyleSheet } from 'react-native'

import { useTheme } from '../../contexts/theme'
import { testIdWithKey } from '../../utils/testable'

const { height } = Dimensions.get('window')

const LoadingModal: React.FC = () => {
  const { ColorPallet, Assets } = useTheme()
  const styles = StyleSheet.create({
    container: {
      minHeight: height,
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: ColorPallet.brand.primary,
    },
  })

  return (
    <Modal visible={true} transparent={true} testID={testIdWithKey('LoadingModalScreen')}>
      <SafeAreaView style={[styles.container]}>
        <Image source={Assets.img.logoLarge} />
        <ActivityIndicator color={ColorPallet.grayscale.white} testID={testIdWithKey('LoadingActivityIndicator')} />
      </SafeAreaView>
    </Modal>
  )
}

export default LoadingModal
