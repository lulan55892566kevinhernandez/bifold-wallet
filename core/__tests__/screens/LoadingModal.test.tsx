import { render } from '@testing-library/react-native'
import React from 'react'

import LoadingModal from '../../App/components/modals/LoadingModal'
import { testIdWithKey } from '../../App/utils/testable'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

describe('displays loading screen', () => {
  test('renders correctly', () => {
    const tree = render(<LoadingModal />)

    expect(tree).toMatchSnapshot()
  })

  test('contains testIDs', () => {
    const tree = render(<LoadingModal />)

    const loadingActivityIndicatorID = tree.getByTestId(testIdWithKey('LoadingActivityIndicator'))

    expect(loadingActivityIndicatorID).not.toBeNull()
  })
})
