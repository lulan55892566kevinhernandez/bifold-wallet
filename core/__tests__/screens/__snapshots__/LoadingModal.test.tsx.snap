// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays loading screen renders correctly 1`] = `
<Modal
  hardwareAccelerated={false}
  testID="com.ariesbifold:id/LoadingModalScreen"
  transparent={true}
  visible={true}
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#42803E",
          "flexDirection": "column",
          "justifyContent": "center",
          "minHeight": 1334,
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "default": "",
        }
      }
    />
    <ActivityIndicator
      color="#FFFFFF"
      testID="com.ariesbifold:id/LoadingActivityIndicator"
    />
  </RCTSafeAreaView>
</Modal>
`;
